File: 11_动态规划/力扣加加/数位dp/357. 计算各个位数不同的数字个数.py
from functools import lru_cache


@lru_cache(None)
def cal(upper: int) -> int:
    @lru_cache(None)
    def dfs(pos: int, visited: int, hasLeadingZero: bool, isLimit: bool) -> int:
        """当前在第pos位，已经用了visited，hasLeadingZero表示有前导0，isLimit表示是否贴合上界"""
        if pos == 0:
            return 1

        res = 0
        up = nums[pos - 1] if isLimit else 9
        for cur in range(up + 1):
            if visited & (1 << cur):
                continue
            if hasLeadingZero and cur == 0:
                res += dfs(pos - 1, visited, True, (isLimit and cur == up))
            else:
                res += dfs(pos - 1, visited | (1 << cur), False, (isLimit and cur == up))
        return res

    nums = []
    while upper:
        div, mod = divmod(upper, 10)
        nums.append(mod)
        upper = div
    return dfs(len(nums), 0, True, True)


class Solution:
    def countNumbersWithUniqueDigits(self, n: int) -> int:
        if n >= 11:
            return 0
        return cal(10 ** n - 1)

File: 11_动态规划/力扣加加/数位dp/357. 计算各个位数不同的数字个数.ts
/**
 * @param {number} n 计算各位数字都不同的数字 x 的个数 其中 0 ≤ x < 10n
 * @return {number}
 * @description
 * n位数 n-1位数 ... 1位数的情况
 */
const countNumbersWithUniqueDigits = function (n: number): number {
  if (n === 0) return 1
  if (n === 1) return 10
  if (n >= 11) return 0
  // 最高位9中选法(1-9) 其余n-1位有 9 8 7 ...
  let res = 9
  for (let i = 0; i < n - 1; i++) {
    res *= 9 - i
  }
  return res + countNumbersWithUniqueDigits(n - 1)
}

// 91
console.log(countNumbersWithUniqueDigits(2))
// 答案应为除去 11,22,33,44,55,66,77,88,99 外，在 [0,100) 区间内的所有数字。
